digital_land/package/sqlite.py

```diff
@@ -46,18 +49,100 @@ def spatialite(self, path=None):
                     path = "/usr/lib/x86_64-linux-gnu/mod_spatialite.so"
         self._spatialite = path
 
-    def connect(self):
+    def connect(self, optimised=False):
         logging.debug(f"sqlite3 connect {self.path}")
-        self.connection = sqlite3.connect(self.path)
+        self.connection = None
+
+        if optimised:
+            self.connection = sqlite3.connect(self.path, cached_statements=0)
+        else:
+            self.connection = sqlite3.connect(self.path)
 
         if self._spatialite:
             self.connection.enable_load_extension(True)
             self.connection.load_extension(self._spatialite)
 
     def disconnect(self):
         logging.debug("sqlite3 disconnect")
+        if self.cursor:
+            self.cursor.close()
+            self.cursor = None
+
         self.connection.close()
 
+    def create_cursor(self, optimised=False):
+        if self.cursor:
+            self.cursor.close()
+            self.cursor = None
+
+        self.cursor = self.connection.cursor()
+        if optimised:
+            self.cursor.execute("PRAGMA synchronous = OFF")
```

Comment:
PARGMA synchronous = OFF and PRAGMA journal_mode = OFF were already being used. If this code is being used outside of the command (which is it) those processes will become less efficient as they're no longer set. 

Check the digital-land-builder repo for another example of where the sqlite package is being used.  by applying this change I imagine those processes would slow down.

Reply:
Ah ok.
The limit of my visibility was the tests.
I looked for other uses in dl-python, but that was it. So apologies for the assumption here.

The intent was that the DDL statements would not use these flags, because it shouldn't be used for these types of statements.
Maybe the answer is to create a dedicated DDL cursor. Then the general create cursor function can have them in as before.

Reply:
Yeah I think that the cursor needs to at least contain what it did before, even if it's an assumption that its optimised. I would also question whether the begin statement should be in the cursor definition. From my perspective the cursor could be used with or without it. It's the individual function for the query which should decide to use it.

Reply:
We're agreed on the first bit.
I'll include that.

Not sure about the second bit, to be honest.
If the cursor client was in charge of the query they run, I would agree with you.
However, the code uses predefined queries in functions and if, as is the case here, the predefined queries are to perform in a uniformly optimal way (i.e., batch commits whenever possible, prepared statements, etc.), then we shouldn't leave the inclusion of the transactional marker to chance.
For example., the DatasetPackage cursor client is only concerned with calling the right sequence of functions and passing the appropriate data. It has no knowledge of whether these should be transactional or not.
I feel your suggestion requires the cursor client to have knowledge of the implementation of the function.